backend/src/main/kotlin/cz/loono/backend/api/service/ExaminationRecordService.kt

```diff
@@ -1,139 +1,84 @@
+                )
+            }
         }
     }
 
-    private fun deleteObsoleteRecords(currentRecords: List<ExaminationRecord>, desiredRecordTypes: List<String>) {
-        val obsoleteTypes = currentRecords.map { it.type } subtract desiredRecordTypes
-        val removals = currentRecords.filter { record ->
-            record.type in obsoleteTypes
-        }
-        examinationRecordRepository.deleteAll(removals)
+    private fun findAccount(uid: String): Account {
+        return accountRepository.findByUid(uid) ?: throw LoonoBackendException(
+            HttpStatus.NOT_FOUND, "Account not found"
+        )
     }
 
-    @Transactional(rollbackFor = [Exception::class])
-    fun completeExamination(
-        uid: String,
-        type: String,
-        completionDate: LocalDate?
-    ): List<ExaminationRecord> {
-        @Suppress("NAME_SHADOWING")
-        val completionDate = completionDate?.withDayOfMonth(1) ?: LocalDate.now().withDayOfMonth(1)
+    private fun changeState(examId: Long, uid: String, state: ExaminationStatusDto): ExaminationRecordDto {
+        val account = findAccount(uid)
 
-        if (completionDate.isAfter(LocalDate.now().withDayOfMonth(1))) {
-            logger.error("Completion date is in the future, this should have been caught by input validator!")
-            throw IllegalArgumentException("Examination Completion must not be in the future.")
-        }
-
-        val account = accountRepository.findByUid(uid)
-        if (account == null) {
-            logger.error("Account does not exist, it should have been created by the account interceptor.")
-            throw IllegalStateException("Tried to complete an examination for uid: $uid but no such account exists.")
-        }
-        val currentRecord = account
-            .withAllRecords()
-            .examinationRecords
-            .first { it.type == type }
+        val exam = examinationRecordRepository.findByIdAndAccount(examId, account)
+        exam.status = state
 
-        var updatedRecord = currentRecord
-
-        if (currentRecord.lastVisit == null || currentRecord.lastVisit.isBefore(completionDate)) {
-            updatedRecord = currentRecord.copy(lastVisit = completionDate)
-        }
-
-        // TODO call Gamification Feature to award points if user is eligible
-
-        accountRepository.save(account.withReplacedRecord(updatedRecord))
-        return getOrCreateRecords(uid)
+        return examinationRecordRepository.save(exam).toExaminationRecordDto()
     }
 
-    fun Account.withReplacedRecord(newRecord: ExaminationRecord): Account {
-        val updatedRecords: List<ExaminationRecord> = examinationRecords
-            .toMutableList()
-            .apply {
-                retainAll { it.type != newRecord.type }
-                add(newRecord)
-            }
-
-        return copy(examinationRecords = updatedRecords)
+    fun ExaminationRecord.toExaminationRecordDto(): ExaminationRecordDto {
```

Comment:
Again optional expression:
```
    fun ExaminationRecord.toExaminationRecordDto(): ExaminationRecordDto =
         ExaminationRecordDto(
            id = id,
            type = type,
            date = date,
            firstExam = firstExam,
            status = status
        )
```

Btw, maybe add this function directly to `ExaminationRecord` instead of an extension function, since we have an access to `ExaminationRecord` class?

Reply:
agree with expression. I decided to insert the function here because of the context with the service and transformation fot the API.